src/mongoose_rabbit_worker.erl

```diff
@@ -0,0 +1,298 @@
+%%==============================================================================
+%% Copyright 2018 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc
+%% Generic RabbitMQ worker. The module is responsible for interacting with
+%% RabbitMQ server. Worker holds its own AMQP connection and channel in its
+%% state.
+%% @end
+%%==============================================================================
+
+-module(mongoose_rabbit_worker).
+-author('kacper.mentel@erlang-solutions.com').
+
+-include_lib("mongooseim/include/mongoose.hrl").
+
+-behaviour(gen_server).
+
+-export([start_link/0, list_metrics/0]).
+
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2]).
+
+-type state() :: #{amqp_client_opts := mongoose_amqp:network_params(),
```

Comment:
This type definition is most probably incorrect. Try adding spec to `init` function. I'm pretty sure all of these have to be defined with `=>`, not `:=`.

Reply:
Here is what I've found in Erlang docs:

```
Association :: Type := Type                  %% denotes a mandatory association
             | Type => Type                  %% denotes an optional association

```

Reply:
Now I see, you're right :)

The connection is done asynchronously to start workers faster.

Reply:
I think we have a race condition then. What if a publish request arrives before the init is complete?

These pools will usually be started on MongooseIM startup, so it's not that important to save like how much? 1s? 2s? Other pool workers are initialised synchronously anyway.